R/regional_epinow.R

```diff
@@ -0,0 +1,303 @@
+#' @param non_zero_points Numeric, the minimum number of time points with non-zero cases in a region required for
+#' that region to be evaluated. Defaults to 2.
+#' @param output A character vector of optional output to return. Supported options are the individual regional estimates
+#' ("regions"),  samples ("samples"), plots ("plots"), timing ("timing"), the stan fit of the underlying model ("fit"), and an 
+#' overall summary across regions ("summary"). 
+#' The default is to return samples and plots alongside summarised estimates and summary statistics. This argument uses partial 
+#' matching so for example passing "sam" will lead to samples being reported.
+#' @param summary_args A list of arguments passed to `regional_summary`. See the `regional_summary` documentation for details.
+#' @param ... Pass additional arguments to `epinow`. See the documentation for `epinow` for details.
+#' @inheritParams epinow
+#' @inheritParams regional_summary
+#' @return A list of output stratified at the top level into regional output and across region output summary output
+#' @export
+#' @importFrom future.apply future_lapply
+#' @importFrom data.table as.data.table setDT copy setorder
+#' @importFrom purrr safely map compact keep
+#' @importFrom futile.logger flog.info flog.warn flog.trace
+#' @importFrom R.utils withTimeout
+#' @importFrom rlang cnd_muffle
+#' @importFrom progressr with_progress progressor
+#' @examples
+#'  \donttest{
+#' # construct example distributions
+#' generation_time <- get_generation_time(disease = "SARS-CoV-2", source = "ganyani")
+#' incubation_period <- get_incubation_period(disease = "SARS-CoV-2", source = "lauer")
+#' reporting_delay <- list(mean = log(10), mean_sd = log(2),
+#'                         sd = log(2), sd_sd = log(1.1), max = 30)
+#'                         
+#' # uses example case vector
+#' cases <- EpiNow2::example_confirmed[1:40]
+#' cases <- data.table::rbindlist(list(
+#'   data.table::copy(cases)[, region := "testland"],
+#'   cases[, region := "realland"]))
+#' 
+#' # setup logging
+#' setup_logging(file = paste0(tempdir(), "/info.log"), 
+#'               name = "EpiNow2.epinow")
+#'                 
+#' # run epinow across multiple regions and generate summaries
+#' # samples and warmup have been reduced for this example
+#' out <- regional_epinow(reported_cases = cases, 
+#'                        samples = 100,
+#'                        generation_time = generation_time,
+#'                        delays = list(incubation_period, reporting_delay),
+#'                        stan_args = list(warmup = 100, 
+#'                                         cores = ifelse(interactive(), 4, 1)))
+#'}
+regional_epinow <- function(reported_cases, 
+                            target_folder = NULL, 
+                            target_date,
+                            non_zero_points = 2, 
+                            return_output = TRUE,
```

Comment:
Yes, good point. The reason to have this is the edge cases where you might want to both return and save output at the same time. Just having a think about whether that is something you might actually ever want to do.

For `epinow` some kind of return control is needed so that `regional_summary` can get the data it needs without looking at the disk regardless of the status of `target_folder`. This has proven a more stable way of generating summary measures in practice as reading from disk can mean accidentally loading old estimates.

Reply:
Gotcha! That makes sense to me. Thanks!

Reply:
Latest commit has updated so that `return_output = FALSE` and is changed internally to `TRUE` if the target directories are missing (with logging saying this is happening). Means that the behavior the user probably wants is default and the edge case is non-default.